dhall/src/Dhall/Import.hs

```diff
@@ -390,29 +390,23 @@ instance Canonicalize Import where
     canonicalize (Import importHashed importMode) =
         Import (canonicalize importHashed) importMode
 
-toHeaders
-  :: Text
-  -> Text
-  -> Expr s a
-  -> Maybe [(CI Data.ByteString.ByteString, Data.ByteString.ByteString)]
-toHeaders key₀ key₁ (ListLit _ hs) = do
-    hs' <- mapM (toHeader key₀ key₁) hs
+toHeaders :: Expr s a -> Maybe [HTTPHeader]
+toHeaders (ListLit _ hs) = do
+    hs' <- mapM toHeader hs
     return (Data.Foldable.toList hs')
-toHeaders _ _ _ = do
+toHeaders _ = do
     empty
 
-toHeader
-  :: Text
-  -> Text
-  -> Expr s a
-  -> Maybe (CI Data.ByteString.ByteString, Data.ByteString.ByteString)
-toHeader key₀ key₁ (RecordLit m) = do
-    TextLit (Chunks [] keyText  ) <- Dhall.Map.lookup key₀ m
-    TextLit (Chunks [] valueText) <- Dhall.Map.lookup key₁ m
+toHeader :: Expr s a -> Maybe HTTPHeader
+toHeader (RecordLit m) = do
+    TextLit (Chunks [] keyText  ) <-
+        Dhall.Map.lookup "header" m <|> Dhall.Map.lookup "mapKey" m
```

Comment:
This is not exactly the same semantics as the standard since it permits `List { header : Text, mapValue : Text }`

You would need to do something like:

```haskell
toHeader (RecordLit m) = do
    let alternative0 = do
            TextLit (Chunks [] keyText) <- Dhall.Map.lookup "header" m
            TextLit (Chunks [] keyValue) <- Dhall.Map.lookup "value" m
            return (keyText, keyValue)

    let alternative1 = do
            TextLit (Chunks [] keyText) <- Dhall.Map.lookup "mapKey" m
            TextLit (Chunks [] keyValue) <- Dhall.Map.lookup "mapValue" m
            return (keyText, keyValue)

    (keyText, valueText) <- alternative0 <|> alternative1

    …
```

Reply:
That the header expression is of a valid type (so either `List { header : Text, value Text}` or `List { mapKey : Text, mapValue Text}`) is one of the invariants of the `Chained` type. In the end we call `toHeader` with a header from a `Chained` remote import, so that should be alright.

I guess I should make that more explicit in the comment describing the `Chained` type.

Reply:
I added a comment to clarify these assumptions.